### PR TITLE
Writing prompts block: load prompts when user not connected to WP.com

### PR DESCRIPTION
**Pull request by creativecoder**
## Proposed changes:

Fixes wpcom/v3/blogging-prompts endpoint so that users not connected to WP.com can still retrieve prompts.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

PT: pe7F0s-uo-p2

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

* Sandbox public-api.wordpress.com and apply the changes from this PR to your sandbox (`bin/jetpack-downloader test jetpack fix/blogging-prompt-wo-wpcom-user`)
* Make sure your Jetpack site is connected to WP.com and beta blocks are enabled (`define( 'JETPACK_BETA_BLOCKS', true );`)
* Created a new editor user on your site and login as that user.
* Create a new post and insert a writing prompt block. The prompt should load without error.

### CODE REVIEW

**Review by jeherve — COMMENTED** [review]
I'm afraid I don't fully understand the user vs. blog request, and why that is needed. Is there some discussion on P2 or Slack where I could learn more about this?

I'll let someone more knowledgeable than me with the API look at this. @fgiannar Do you think you could take a look?

Thank you!

**Comment by creativecoder** [comment]
> I'm afraid I don't fully understand the user vs. blog request, and why that is needed. Is there some discussion on P2 or Slack where I could learn more about this?

The as user request is preferred, since the response includes information that's user specific, mainly if they have answered the prompt before. We're not using that information now, but I imagine it being used in the future to exclude prompts that have already been answered by the user.

The as blog request is added here as a fallback. If you open the editor as a user on the site who isn't connected to WP.com, the block will show an error because you can't access the blogging prompts for the site (as your user authentication fails against the WP.com API).

**Comment by jeherve** [comment]
> The as user request is preferred, since the response includes information that's user specific, mainly if they have answered the prompt before. We're not using that information now, but I imagine it being used in the future to exclude prompts that have already been answered by the user.

Ah, that makes sense. In this case, wouldn't we want to allow folks to answer the same prompt on multiple sites, if they do have multiple blogs? I could imagine someone with a very public blog and a personal blog, with different audiences, and wanting to offer different answers on each blog.

**Comment by creativecoder** [comment]
> In this case, wouldn't we want to allow folks to answer the same prompt on multiple sites, if they do have multiple blogs? 

The blogging-prompts endpoints are site specific when served from wpcom (`wpcom/v3/sites/{site_id}/blogging-prompts`), so it would support updating the `answered` property to take blog_id into account (though it looks like it isn't right now--see `answers.php` in wpcom)

**Review by eoigal — DISMISSED** [review]
Works as advertised 👍

File: projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v3-endpoint-blogging-prompts.php [hunk 1]
```diff
@@ -370,15 +380,16 @@ public function get_item_schema() {
 	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
 	 */
 	public function permissions_check() {
```
Comment (fgiannar) [inline projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v3-endpoint-blogging-prompts.php:399]: Maybe we could use a slightly simpler approach here, depending on where the code is running:

Case 1: User hitting the endpoint hosted on their Jetpack site, from their Jetpack site, or hitting the endpoint hosted on WPCOM, from their WPCOM site. 

```
if ( current_user_can( 'edit_posts' ) ) {
	return true;
}
```

Case 2: Jetpack hitting the endpoint hosted on WPCOM, from a Jetpack site with a blog token.

```
if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
	if ( is_jetpack_site( $blog_id ) ) {
		if ( ! class_exists( 'WPCOM_REST_API_V2_Endpoint_Jetpack_Auth' ) ) {
			require_once dirname( __DIR__ ) . '/rest-api-plugins/endpoints/jetpack-auth.php';
		}

		$jp_auth_endpoint = new WPCOM_REST_API_V2_Endpoint_Jetpack_Auth();
		if ( true === $jp_auth_endpoint->is_jetpack_authorized_for_site() ) {
			return true;
		}
	}
}
```

We have already prior art for this approach in the [tweetstorm helper endpoints](https://github.com/Automattic/jetpack/blob/b6bfcca89a300b5e814a84a52f45aa98b15e231a/projects/plugins/jetpack/_inc/lib/class-jetpack-tweetstorm-helper.php#L220)

This way we also don't need the `WPCOM_REST_API_V2_Jetpack_Auth_Trait` trait.

How would you feel about this approach?

cc @jeherve

Comment (creativecoder) [reply projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v3-endpoint-blogging-prompts.php:399]: That works for me, and I like that it's simpler!

**Comment by creativecoder** [comment]
PR and testing instructions are now updated to use @fgiannar 's suggestions.

**Review by fgiannar — APPROVED** [review]
Many thanks for implementing the suggested changes!

I've tested this both on a Jetpack site as a connected admin and non connected editor and it works as described.

I've also tested this on a Simple site and can confirm it works great too!

Well done!

**Comment by creativecoder** [comment]
FYI that I created a follow-up to this PR that resolves the API tests not passing on wpcom.

https://github.com/Automattic/jetpack/pull/29547

**Pull request merged**